### ISSUE TITLE
Typein and Modulator Bugs

### ISSUE THREAD

**Issue by baconpaul** [surge-synthesizer/surge#2229]
Leaving this one open while I think about that spacing issue, and to see if other bugs come in.

Typein and Modulator Bugs

- [x] Sin HiCut is in Semitons on my mac
- [x] Oh I didn't test the Decibel case carefully enough. Thank you @Jacky-Ligon - that's the send problem
- [ ] Agree with spacing comments. Need to think about how to do this better. Pushed some padding for now.
- [x] Agree i need to fix FMRatio - I must have missed a custom case.
- [x] @mkruselj the "8123 seconds" or whatever gets subbed with "Forever" in the value and should get similarly subbed in the modulation.
- [x] Specious "+" on -ve values in the full linear case
- [x] Mixer shows old school. Generally "I skipped Decibels"
- [x] Sine Os High Cut displays differently then Low cut and with much higher values then anything else.
- [x] +/- -> -/+ on bipolars
- [x] M1 Ratio is custom and doesn't do the 5-way thing
- [x] M1 Ratio is custom and doesn't support types properly

**Comment by Jacky-Ligon** [comment]
**Surge-NIGHTLY-2020-06-28-0afefa8 | Win-10, VST3i**

**Sine Osc Typein Issues**
Assign Velocity or Keytrack, to Low or High Cut filters and move blue handle to right.
RC and typein 500 Hz.
Check result, which shows 513.750 Hz (and apparent rounding error).

![image](https://user-images.githubusercontent.com/66483816/85959600-637d0b80-b96b-11ea-89cd-6a98529da3d6.png)


**Modulation Of Mixer Levels Typein Issue**
Pull down Mixer Channel 1 to -inf dB.
Assign a Velocity, ENV, LFO, or SLFO modulation-source to positively modulate the Mixer Channel in from -INF to 0 dB.
As one moves the Mixer channel slider up, the leftmost value shows -INF, and the rightmost 0 dB.
RC on the blue handle and try to type in a new value for the rightmost value shown when setting the modulation initial range, and one will see that it is actually the leftmost value that is displayed for editing, rather than the right.
Try a type in value of -24 dB, and one will observe that the blue handle of the Mixer channel, rather than being set to -24 dB, is rather set to -INF.
In this scenario, it appears that the typein values are inverted, and is not possible to typein a value less than 0 dB without setting the control to -INF.

![LICEcap](https://user-images.githubusercontent.com/66483816/85959725-0d5c9800-b96c-11ea-883f-522514b40dac.gif)


**Surge-NIGHTLY-2020-06-27-4ec1fc6 | Windows-10, VST3i | Typein Issues**
Filter Modulation Typein Issues
RC on a Filter handle, type in 500 Hz for the cutoff frequency.
Assign Velocity to positively modulate the cutoff frequency with a range of 500 Hz, to 5000 Hz.
RC on the blue handle, and typein 5000, then RC again to check the value...
Rather than the expected 5000 Hz, 5500 Hz is displayed instead.
This rounding error (?) seems to manifest with all of the other modulation-sources as well.

![LICEcap](https://user-images.githubusercontent.com/66483816/85959806-9e337380-b96c-11ea-9953-6fc764b2dc14.gif)


**Send FX and Return FX Modulation Typein Issues**
RC on any Send or Return FX slider, and typein -12 dB to set the initial center value of the control.
Assign a S-LFO to the control for bipolar modulation.
Move the blue modulation range handle to the right where the readout shows approximately -6 dB for the rightmost value, then RC and typein -6 dB, and observe that the blue handle swaps sides of the control center.
RC on the blue modulation range control, and notice that the displayed values show -7.12 ... -18 dB.

![LICEcap](https://user-images.githubusercontent.com/66483816/85959864-0bdf9f80-b96d-11ea-90f8-ef8bda3da64a.gif)

**Comment by baconpaul** [comment]
Just for posterity: Those reports are all examples of the synth doing what I intended it to. So they weren't coding bugs. They were design-time UX bugs. New push coming soon which shows the data rather differently.

**Comment by VincyZed** [comment]
![image](https://user-images.githubusercontent.com/50145178/85962761-74844780-b980-11ea-96ad-6eef7e7637c6.png)

**Comment by baconpaul** [comment]
That gets fixed by adding a modulationCap to displayInfo for the freqs. Will push with others at end of day if other bugs come in.

**Comment by baconpaul** [comment]
Oh I accidentally pushed the fix to that with #2238

**Comment by Jacky-Ligon** [comment]
**Surge-NIGHTLY-2020-06-29-9c42bd3 | Win-10, VST3i**

**FM3 Oscillator doesn't accept typeins for M1 or M2 Ratios**

![LICEcap](https://user-images.githubusercontent.com/66483816/86041816-39cbef00-ba14-11ea-9193-845d27ce594b.gif)

**Comment by baconpaul** [comment]
Oh thank you! That's a real bug indeed.

**Comment by Jacky-Ligon** [comment]
**Surge-NIGHTLY-2020-06-29-9c42bd3 | Win-10, VST3i**

**modulationCap issue for all oscillators with High Cut filters**

Overlapping values and label when mapping Velocity modulation to High Cut filters, and ambiguous results with typeins of cutoff frequencies.

![LICEcap](https://user-images.githubusercontent.com/66483816/86045377-7f3eeb00-ba19-11ea-9ea4-4f4be8cf25dd.gif)

**Comment by baconpaul** [comment]
@Jacky-Ligon after a long conversation on slack, we/ve made some more changes simplifying the UI in the base case. Just GYI. There's a new nightly. Things are moving quickly. Apologies. This is hard!

**Comment by Jacky-Ligon** [comment]
Confirming that the above is fixed in Surge-NIGHTLY-2020-06-29-b1c054b

**Comment by Jacky-Ligon** [comment]
**Surge-NIGHTLY-2020-06-29-b1c054b | Win-10, VST3i

Modulated FEG and AEG release times can be set up to 8,191 seconds.**

![LICEcap](https://user-images.githubusercontent.com/66483816/86050093-460a7900-ba21-11ea-8228-122d0712d7b6.gif)

**Comment by Jacky-Ligon** [comment]
**Surge-NIGHTLY-2020-06-29-b1c054b | Win-10, VST3i**

**Unexpected results when modulating FX Sends and Returns with S-LFO**

There appears to be no correlation between slider-values, typein values, and values displayed in context menu after typeins.

![LICEcap](https://user-images.githubusercontent.com/66483816/86052346-d72f1f00-ba24-11ea-82c9-3c9de8672c8b.gif)

**Comment by mkruselj** [comment]
> Modulated FEG and AEG release times can be set up to 8,191 seconds.

Not a bug. They can really go on for that long.

> There appears to be no correlation between slider-values, typein values, and values displayed in context menu after typeins, although the Returns make a little more sense than the Sends.

Another issue here is that these don't show the information in the same way that other parameters do - there is no 3rd line with deltas when "Modulation popups show bounds" option is enabled. But also yes I see the discrepancy between the delta shown in the menu/popup and the value in typein.

**Comment by mkruselj** [comment]
![image](https://user-images.githubusercontent.com/2393720/86060237-a1aa2600-ba64-11ea-835f-de2748938d11.png)

Could probably give more space to the base value in the middle.

**Comment by baconpaul** [comment]
- [ ] Oh I didn't test the Decibel case carefully enough. Thank you @Jacky-Ligon - that's the send problem
- [ ] Agree with spacing comments. 
- [ ] Agree i need to fix FMRatio - I must have missed a custom case.
- [ ] @mkruselj the "8123 seconds" or whatever gets subbed with "Forever" in the value and should get similarly subbed in the modulation.

Thank you all.

**Comment by K0rrid0r** [comment]
![116](https://user-images.githubusercontent.com/56724286/86077433-412edf00-ba8c-11ea-8933-0cbf56e14a54.gif)
"+-" is shown at negative values in oscillator window.

![117](https://user-images.githubusercontent.com/56724286/86077533-6ae80600-ba8c-11ea-9b27-4bcd12b1022c.gif)
differences in display in the mixer, should be shown as Gain info is.

![118](https://user-images.githubusercontent.com/56724286/86077714-c4e8cb80-ba8c-11ea-82a7-6db501c14575.gif)
Phase / Shuffle
Amplitude
Deform
shows "+-"

![119](https://user-images.githubusercontent.com/56724286/86077850-0da08480-ba8d-11ea-876f-4b427022f9eb.gif)
S shows "+-"
R overlaps the numbers. far to many decimals here.

![120](https://user-images.githubusercontent.com/56724286/86077955-3fb1e680-ba8d-11ea-9977-5e7a268a2677.gif)
Filter Keytracking displays "+-"
Waveshaper displays "+-"

![121](https://user-images.githubusercontent.com/56724286/86078035-77209300-ba8d-11ea-871a-69df97f1532d.gif)
Polymode displays "+-"

![122](https://user-images.githubusercontent.com/56724286/86078170-cf579500-ba8d-11ea-8f2b-591312860152.gif)
Resonance in both filters displays "+-"
F1 F2 slider displays "+-"

![123](https://user-images.githubusercontent.com/56724286/86078325-2c534b00-ba8e-11ea-8611-193be1f54c8d.gif)
F1 F2 slider displays "+-"
Amp EG S displays "+-"
Gain displays "+-"

![124](https://user-images.githubusercontent.com/56724286/86078498-9370ff80-ba8e-11ea-843d-afd4ca4b0509.gif)
FM Depth displays "+-"
Feedback displays "+-"
Width displays "+-"
Pan displays "+-"
Send 1  displays other kind of window that should be the same as the other ones.
Send 2  displays other kind of window that should be the same as the other ones.

![125](https://user-images.githubusercontent.com/56724286/86078746-11350b00-ba8f-11ea-978a-5b10866732c6.gif)
Master volume displays "+-"
Send 1  displays other kind of window that should be the same as the other ones.
Send 2  displays other kind of window that should be the same as the other ones.

**Comment by baconpaul** [comment]
This is amazingly useful feedback @K0rrid0r and @Jacky-Ligon both - thank you very much

**Comment by baconpaul** [comment]
I have consolidated this feedback at the top of the issue and will work from that list. Thank you.

**Comment by baconpaul** [comment]
@K0rrid0r in my test, hi cut and low cut show the exact same. all the frequencies have a cap though. What's the bug exactly?

**Comment by K0rrid0r** [comment]
Look at the difference on display, is that really correct?

![136](https://user-images.githubusercontent.com/56724286/86123723-43b82580-bada-11ea-9ff2-84571a47da7e.gif)

**Comment by mkruselj** [comment]
Yes, that is indeed correct.

**Comment by K0rrid0r** [comment]
I see it now. i guess too much info about everything at this moment.

**Comment by baconpaul** [comment]
well also some pretty poor spacing makes it hard to read.

**Comment by K0rrid0r** [comment]
![145](https://user-images.githubusercontent.com/56724286/86193297-db099100-bb4b-11ea-8a56-e45ac698b25f.gif)
The two F1 F2 filter sections displays different values where as the ones in waveshaper are correct with +/- 200
F1 and F2 in Filter EG displays +/- 192

**Comment by baconpaul** [comment]
So @K0rrid0r each of these are on purpose. It doesn't mean they are right but I'm not quite sure they are bugs. Let me explain

1. We cap all gains at -192db. The mixer gain has a range which goes down that low so if you modulate below -192db you end up getting no extra modulation. Since the overarching gain has a range of +/-48db you can fully modulate down by -96 and not hit the -192; that's why it shows

2. The reason A, D, and R have their lower bound at zero is they are times. You can't have a negative time attack. So it's the same thing - we are running into a bound and showing it properly.

3. The KeyTrack one is more irritating. The display is correct *but* the keytrack F1/F2 is in units of percent, and the Filter F1/F2 is in units of semitones. The reason for this (in the code) is because the keytrack is multiplied by key number to get the offset (which is in pitches), but the envelope is not. So this is "right but confusing as all heck". However the display does show semitones for the F1/F2 in env and percent for the F1/F2 in keytrack, so the UI correctly shows this. It is very strange though that the two items labeled the same mean different things (in one case, how much to scale the keytrack to apply to the cutoff, and in the second case, how many pitches to multiply the envelope by to apply to the cutoff).

So I'm not quite sure what to fix. I think the first is the only one where we could do anything (the thing to do would be to not cap the mod distances) and I'm happy to do that, but I think the second two are correct.

Thoughts?

**Comment by K0rrid0r** [comment]
thanks for clarification. lets see what others may think, i was just used to see negative modulations that's all :)

**Comment by Jacky-Ligon** [comment]
**Surge-NIGHTLY-2020-07-01-c84b66a | Win-10, VST3i**

**Sends and Returns FX Modulation Typeins scale tracks and handles differently with the same input values**

![LICEcap](https://user-images.githubusercontent.com/66483816/86293061-eb5b5380-bbbf-11ea-87b3-5fa152d874b3.gif)

**Comment by baconpaul** [comment]
Oh @mkruselj didn't you find some place where that control was manually adjusting ranges in Parameter space? That ugly thing we found that makes the send have a different max than the return?

**Comment by baconpaul** [comment]
Anyway @Jacky-Ligon the reason for this is (on purpose) the send and return have different maxes so they will scale differently.

Why exactly that was done on purpose, and the gross way it was implemented, is an interesting question. But this isn't a bug, it's a feature. Which we may want to fix! But the rendering is I think correct.

**Comment by mkruselj** [comment]
> Oh @mkruselj didn't you find some place where that control was manually adjusting ranges in Parameter space? That ugly thing we found that makes the send have a different max than the return?

Yes I did, now it has its own ctype (ct_sendlevel). Returns use ct_amplitude IIRC.

**Comment by mkruselj** [comment]
Also, I think it's good that modulation amounts are capped. If your osc volume is already at -inf (which means complete silence), modulating that with a negative value is... more negative infinity? Makes no sense. So showing 0 dB is good here. Ditto for envelope times not going below 0.000 s. I don't think it should change.

**Comment by baconpaul** [comment]
Yeah on @Jacky-Ligon thing I did a double check to make sure we had handled all the max/min differently. We had and the sliders are consistent with their ranges, even though their ranges are different.

So my instinct is: We have this correct. There's some *parameters* which are weird in surge (the differences in keytrack meaning; the differences in range) but the UI and type-ins correctly represent those strangenesses.

I'm going to close this ticket as resolved the. We can easily reopen it in the beta period if we find something.

And most importantly: @Jacky-Ligon and @K0rrid0r thank you so much for your work with @mkruselj and me on designing and testing this feature. This was a tricky one and I'm super happy with where we landed.
